index.js

```diff
@@ -85,27 +87,44 @@ module.exports = function (Adapter) {
           return error ? reject(error) : resolve(files)
         })
       })
-    ).then(function (files) {
-      return Promise.all(map(files, function (file) {
-        return new Promise(function (resolve, reject) {
-          var filePath = path.join(dbPath, type, '' + file)
+    ).then(async function (files) {
+      const allThePromises = []
+      let readsInFlight = 0
+      const iterator = files[Symbol.iterator]()
 
-          fs.readFile(filePath, function (error, buffer) {
-            var record
+      while (allThePromises.length < files.length) {
+        if(readsInFlight >= concurrentReads) {
+          // back off
+          await pause(10)
```

Comment:
hmm this is an unusual way to do it, and quite finicky. some situations where this is sub-optimal:
- reads are super fast and result in idling for the majority of 10ms
- reads take 10ms longer than they should because there's just 1 more file to read

ideally the next promise starts immediately when a previous one is resolved.

there are some good examples in the comments of [this gist](https://gist.github.com/jcouyang/632709f30e12a7879a73e9e132c0d56b) or even a [package](https://www.npmjs.com/package/p-limit) though i'd prefer not to import a package for a trivial amount of code.

Reply:
alright, I rolled my own, and perf was not as good as the 10ms debounce impl.. a couple seconds slower at between 8 and 9 seconds.  of course, a different dataset may yield different results, but still.

I also gave the gist a shot and it was even worse: nearly 13 seconds

p-limit, however, was faster than all these impls...  just shy of 5.5 seconds consistently. looking at the lib code, it's hard to discern why the perf is better than what I hand-rolled.  perf isn't everything, but I'll continue to experiment with optimizations

I agree with you; I am also loathe to add a dependency for this

Reply:
well, we're not going to do much better than just setting the sleep to `0` just to get an event loop microtask delay

now some interesting results from benchmarking again:

| concurrent read limit | execution time (ms) |
| --- | --- |
|32 | 10632|
|64 | 7076|
|128 | 6706|
|256 | 6865|
|512 | 6880|
|1024 | 6918|
|2048 | 6972|

we now see the cliff at `128`, and degraded performance past that.  so it looks like the magic number if using the `0` back off is `128`

Reply:
alright code is updated.  also made it configurable, added tests, and updated the readme.  ✨

Reply:
interesting... 0 is effectively the same as next tick so i guess that works.